File: packages/diseno_flutter/lib/moviles/foco_de_tutorial.dart
import 'package:flutter/widgets.dart';

import '../tokens/tokens.dart';

/// Aire alrededor del elemento, para que el anillo no lo toque.
const double _holgura = 6;

/// **El velo con un agujero.** Oscurece la pantalla salvo el elemento del que habla el
/// tutorial, y dibuja un anillo alrededor: el foco no se comunica solo con la sombra,
/// porque el oscurecimiento solo no se ve en una pantalla al sol ni con poca vista.
///
/// El agujero **deja pasar el toque** salvo que [bloquea] sea verdadero: un paso que
/// dice «tocá este botón» tiene que dejar tocarlo de verdad.
class FocoDeTutorial extends StatelessWidget {
  const FocoDeTutorial({
    super.key,
    required this.recuadro,
    this.bloquea = false,
    this.alTocarFuera,
  });

  /// `null` mientras se busca el elemento o cuando el paso no señala nada: el velo se
  /// cierra entero y el globo habla de la pantalla.
  final Rect? recuadro;
  final bool bloquea;
  final VoidCallback? alTocarFuera;

  @override
  Widget build(BuildContext context) {
    final t = Tokens.of(context);
    final quieto = MediaQuery.disableAnimationsOf(context);
    final actual = recuadro;
    final pintura = actual == null
        // Sin recuadro no hay nada que animar —y `RectTween` sin destino revienta—:
        // el velo se cierra entero y el globo habla de la pantalla.
        ? CustomPaint(
            painter: _PintorDeFoco(
              recuadro: null,
              velo: Paleta.ink,
              anillo: t.accent,
            ),
            size: Size.infinite,
          )
        : TweenAnimationBuilder<Rect?>(
            tween: RectTween(end: actual),
            duration: quieto
                ? Duration.zero
                : const Duration(milliseconds: 220),
            curve: Curves.easeOutCubic,
            builder: (context, animado, _) => CustomPaint(
              painter: _PintorDeFoco(
                recuadro: animado ?? actual,
                velo: Paleta.ink,
                anillo: t.accent,
              ),
              size: Size.infinite,
            ),
          );
    // Con el hueco tocable, el velo no puede comerse los toques de la app: se ignora
    // el puntero y el agujero queda vivo porque no hay nada encima de él.
    return ExcludeSemantics(
      child: IgnorePointer(
        ignoring: !bloquea,
        child: GestureDetector(
          onTap: alTocarFuera,
          behavior: HitTestBehavior.opaque,
          child: pintura,
        ),
      ),
    );
  }
}

class _PintorDeFoco extends CustomPainter {
  const _PintorDeFoco({
    required this.recuadro,
    required this.velo,
    required this.anillo,
  });

  final Rect? recuadro;
  final Color velo;
  final Color anillo;

  @override
  void paint(Canvas lienzo, Size medida) {
    final pantalla = Offset.zero & medida;
    final pintura = Paint()..color = velo.withValues(alpha: 0.62);
    if (recuadro == null) {
      lienzo.drawRect(pantalla, pintura);
      return;
    }
    final hueco = RRect.fromRectAndRadius(
      recuadro!.inflate(_holgura),
      const Radius.circular(Radios.lg),
    );
    final conAgujero = Path.combine(
      PathOperation.difference,
      Path()..addRect(pantalla),
      Path()..addRRect(hueco),
    );
    lienzo.drawPath(conAgujero, pintura);
    lienzo.drawRRect(
      hueco,
      Paint()
        ..color = anillo
        ..style = PaintingStyle.stroke
        ..strokeWidth = Borde.foco,
    );
  }

  @override
  bool shouldRepaint(_PintorDeFoco anterior) =>
      anterior.recuadro != recuadro ||
      anterior.velo != velo ||
      anterior.anillo != anillo;
}

File: packages/ui/src/foco-de-tutorial/foco-de-tutorial.ts
import { ChangeDetectionStrategy, Component, computed, input, output } from '@angular/core'

/** El recuadro del elemento señalado, en coordenadas de la ventana. */
export interface RecuadroResaltado {
  readonly x: number
  readonly y: number
  readonly ancho: number
  readonly alto: number
}

/**
 * **El velo con un agujero.** Oscurece la pantalla salvo el elemento del que habla el
 * tutorial, y dibuja un anillo alrededor para que se vea también sin color (el foco no
 * se comunica solo con el oscurecimiento: hay un borde).
 *
 * El agujero **deja pasar el ratón** salvo que `bloquea` sea verdadero: un paso que
 * dice «pulsá este botón» tiene que dejar pulsarlo de verdad. El halo es decorativo;
 * cuando el paso bloquea la app, la cortina ofrece una salida accesible.
 */
@Component({
  selector: 'ap-foco-de-tutorial',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    @if (bloquea()) {
      <button type="button" class="cortina" aria-label="Salir del tutorial" (click)="fueraDelFoco.emit()"></button>
    }
    <div class="halo" aria-hidden="true" [class.sin-recuadro]="recuadro() === null" [style.left.px]="caja().x" [style.top.px]="caja().y" [style.width.px]="caja().ancho" [style.height.px]="caja().alto"></div>
  `,
  styles: `
    :host { position: fixed; inset: 0; z-index: 40; pointer-events: none; }
    .cortina { position: absolute; inset: 0; pointer-events: auto; border: 0; background: transparent; cursor: pointer; }
    .halo {
      position: absolute;
      border-radius: var(--r-lg);
      box-shadow: 0 0 0 100vmax color-mix(in srgb, var(--ink) 62%, transparent);
      outline: var(--borde-foco) solid var(--accent);
      outline-offset: var(--borde-desfase);
      transition: left .22s ease, top .22s ease, width .22s ease, height .22s ease;
    }
    .halo.sin-recuadro { outline: 0; }
    @media (prefers-reduced-motion: reduce) { .halo { transition: none; } }
  `,
})
export class FocoDeTutorial {
  /** `null` mientras se busca el elemento o cuando el paso no señala nada. */
  readonly recuadro = input.required<RecuadroResaltado | null>()
  /** Con `true` nadie toca la aplicación; con `false`, solo el agujero es tocable. */
  readonly bloquea = input(true)
  /** Alguien pulsó fuera del foco: el anfitrión decide si eso es «salir». */
  readonly fueraDelFoco = output<void>()

  /** Sin recuadro, el velo se cierra sobre el centro: queda oscuro y sin anillo. */
  protected readonly caja = computed<RecuadroResaltado>(() => {
    const r = this.recuadro()
    if (r !== null) return { x: r.x - HOLGURA, y: r.y - HOLGURA, ancho: r.ancho + HOLGURA * 2, alto: r.alto + HOLGURA * 2 }
    return { x: 0, y: 0, ancho: 0, alto: 0 }
  })
}

/** Aire alrededor del elemento, para que el anillo no lo toque. */
const HOLGURA = 6
